src/librbd/mirror/GroupEnableRequest.cc

```diff
@@ -0,0 +1,824 @@
+
+  if (r < 0) {
+    lderr(m_cct) << "failed to notify mirror group update: " << cpp_strerror(r)
+                 << dendl;
+  }
+
+  close_images();
+}
+
+template <typename I>
+void GroupEnableRequest<I>::close_images() {
+  ldout(m_cct, 10) << dendl;
+
+  auto ctx = create_context_callback<
+    GroupEnableRequest<I>, &GroupEnableRequest<I>::handle_close_images>(this);
+
+  auto gather_ctx = new C_Gather(m_cct, ctx);
+
+  for (auto ictx: m_image_ctxs) {
+    ictx->state->close(gather_ctx->new_sub());
+  }
+
+  gather_ctx->activate();
+}
+
+template <typename I>
+void GroupEnableRequest<I>::handle_close_images(int r) {
+  ldout(m_cct, 10) << "r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(m_cct) << "failed to close images: " << cpp_strerror(r) << dendl;
+    if (m_ret_val == 0) {
+      m_ret_val = r;
+    }
+  }
+
+  if (m_ret_val < 0) {
+    finish(m_ret_val);
+    return;
+  }
+
+  finish(0);
+}
+
+template <typename I>
+void GroupEnableRequest<I>::remove_primary_group_snapshot() {
+  ldout(m_cct, 10) << dendl;
+
+  auto ctx = create_context_callback<
+    GroupEnableRequest<I>,
+    &GroupEnableRequest<I>::handle_remove_primary_group_snapshot>(this);
+
+  auto req = snapshot::RemoveGroupSnapshotRequest<I>::create (m_group_ioctx,
+     m_group_id, &m_group_snap, &m_image_ctxs, ctx);
+
+  req->send();
+}
+
+
+template <typename I>
+void GroupEnableRequest<I>::handle_remove_primary_group_snapshot(int r) {
+  ldout(m_cct, 10) << "r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(m_cct) << "failed to remove mirror group snapshot: " << cpp_strerror(r)
+                 << dendl;
+    close_images();
+    return;
+  }
+  remove_image_states();
+}
+
+template <typename I>
+void GroupEnableRequest<I>::cleanup_mirror_images() {
+  ldout(m_cct, 10) << dendl;
+
+  if (m_cleanup_image_states) {
+    m_mirror_images.clear();
+    get_mirror_images();
+  } else {
+    remove_primary_group_snapshot();
+  }
+}
+
+template <typename I>
+void GroupEnableRequest<I>::disable_mirror_images() {
+  ldout(m_cct, 10) << dendl;
+
+  auto ctx = create_context_callback<
+    GroupEnableRequest<I>,
+    &GroupEnableRequest<I>::handle_disable_mirror_images>(this);
+
+  auto gather_ctx = new C_Gather(m_cct, ctx);
+  for (size_t i = 0; i < m_images.size(); i++) {
+    if (m_mirror_images[i].state != cls::rbd::MIRROR_IMAGE_STATE_DISABLED) {
+      auto req = ImageStateUpdateRequest<I>::create(
+		 m_image_ctxs[i]->md_ctx, m_image_ctxs[i]->id,
+                 cls::rbd::MIRROR_IMAGE_STATE_DISABLING,
+                 m_mirror_images[i], gather_ctx->new_sub());
+     req->send();
+    }
+  }
+  gather_ctx->activate();
+}
+
+template <typename I>
+void GroupEnableRequest<I>::handle_disable_mirror_images(int r) {
+  ldout(m_cct, 10) << "r=" << dendl;
+
+  if (r < 0) {
+    lderr(m_cct) << "failed to cleanup mirror image: " << cpp_strerror(r)
+                 << dendl;
+    close_images();
+    return;
+  }
+
+  remove_primary_group_snapshot();
```

Comment:
Don't we want to remove group member mirror images (remove_image_states()) before removing the primary group snapshot(remove_primary_group_snapshot())?

Reply:
We want to remove the mirror snaps first because if it fails, there needs to be a MirrorImage to indicate that a cleanup is required.
DisableRequest follows the same sequence.